### ISSUE TITLE
Design questions ...

### ISSUE THREAD

**Issue by Jaykul** [tusdotnet/tusdotnet#51]
I have two questions, basically. I suspect they're a matter of opinion, but if you agree with me, I'd be willing to send some PRs ... 

#### First, it seems it might be better if there was an abstract base class for the configuration. 

I saw from [the docs](https://github.com/tusdotnet/tusdotnet/wiki/Configure-tusdotnet) that you must have used an interface at one time, and I so I wondered what made you decide to make `DefaultTusConfiguration` not be an abstract base.  That way, instead of needing to write my whole implementation inline in my Startup class, I could just subclass `BaseTusConfiguration` and override whatever I need to ...

I was able to treat it as a base class, and derive from it, except that because the events are in their own class, I still had to hook up all the logic in the constructor ... 

#### Second, what's the use case for a write-only store?

Separating all of the storage interfaces individually (without a dependency chain), seems to make this awkward to use with a dependency injection model.

Specifically, I want to register the store for use in my configuration, something like this (in ASP.NET Core):

```csharp
services.AddTransient<ITusStore>(provider => 
                  new TusDiskStore(Configuration.GetValue("uploadPath", @"S:\")));
```

However, when the uploads are done, I need to do something with them, which means I need an `ITusReadableStore` ... is it reasonable to register another interface of the TusDiskStore for each feature I need?

Am I missing something, or would it be reasonable that every Store _must_ derive from `ITusStore`, and perhaps even require the implementation of the `GetFileAsync` method?

I'm honestly not sure how your actual implementation is done -- if I provide a Store that's missing interfaces, do you just disable that functionality on the middleware? Could there ever be such a thing as a write-only or read-only store?

**Comment by smatsson** [comment]
Hi @Jaykul,

Good questions! I'll answer them one by one.

"First, it seems it might be better if there was an abstract base class for the configuration."

It sure was an interface back in the day but as the interface didn't quite do anything that the `DefaultTusConfiguration` class already does, it was removed in tusdotnet 2.0. Instead all properties of the `DefaultTusConfiguration` class were made virtual to allow subclassing. Do you have an example of what you want to achieve using an abstract base class that you cannot achieve with the current implementation of `DefaultTusConfiguration`? Making it abstract would require all who use tusdotnet to subclass and make their own implementation which I don't think is feasible. 

"Second, what's the use case for a write-only store?"

I can definitely see where you're coming from here. The first work on tusdotnet started before IoC was a big thing in C# and hence it was not taken into consideration at all. The store's design pattern was created to mimic that of the tus protocol which consists of a mandatory core protocol (represented by `ITusStore`) and several extensions (e.g. `creation`/`ITusCreationStore`, `termination`/`ITusTerminationStore` etc). Since reading files is not part of the protocol it was decided to put this into a separate interface (`ITusReadableStore`).  You are absolutely correct in that tusdotnet will disable features based on the store that is provided. The only interface that is required to be implemented by the store is `ITusStore` as this represents the core protocol. 

Without knowing that much about your use case I would think that you would need to setup `ITusStore` for your configuration of tusdotnet and `ITusReadableStore` to be able to read files from some other place in your code base. I completely understand why this is not optimal. Perhaps adding an extra, empty interface between `TusDiskStore` and all the `ITusXXXStore` interfaces would help? That way one could hook up `ITusDiskStore` to use `TusDiskStore`  in the IoC and register it only once. Since `ITusDiskStore` would inherit from all the other interfaces it should be fine to just fetch it and provide it to the `DefaultTusConfiguration`.

**Comment by smatsson** [comment]
No answer for 30 days. Closing due to inactivity.
